src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1261,63 +1269,101 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
             return cachedItem;
         }
 
+        // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
         RefreshItem refreshItem = getRefreshToken(request);
+        AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
-                && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
+                && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) 
+        {
             Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
-            return refreshToken(callbackHandle, activity, useDialog, request, refreshItem, true);
-        } else {
-            Logger.v(TAG, "Refresh token is not available");
+            try
+            {
+                authResult = getTokenWithRefreshToken(activity, useDialog, request, refreshItem, true);
+            }
+            catch (AuthenticationException authenticationException)
+            {
+                callbackHandle.onError(authenticationException);
+                return null;
+            }
+            
+            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken()))
+            {
+                callbackHandle.onSuccess(authResult);
+                return authResult;
+            }
+        } 
+        
+        // refresh token does not exist or refresh token request failed to give back the token. 
+        // If it's non-silent request, will try to acquire token interactively. 
+        // if it's silent request, will return the AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED.
+        if (refreshItem == null || authResult == null 
+                || (authResult != null && StringExtensions.IsNullOrBlank(authResult.getAccessToken()))) 
+        {
+            Logger.v(TAG, "Refresh token is not available or refresh token request failed to return token.");
```

Comment:
Scenario: signed in to Word->didn't use app for a while and token got expired-> go to Excel and sign in -> go back to Word -> it will check token and will ask for credentials...but it shouldn't